libraries/AP_HAL_Linux/PWM_Sysfs.cpp

```diff
@@ -0,0 +1,168 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <AP_HAL/AP_HAL.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#include "PWM_Sysfs.h"
+
+#include <fcntl.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+using namespace Linux;
+
+static const AP_HAL::HAL& hal = AP_HAL::get_HAL();
+
+#define BASE_PWM_PATH "/sys/class/pwm/pwmchip%d"
```

Comment:
I actually think it's worse to define `BASE_PWM_PATH` with a format specifier... because then you look in the place doing the printf and and number for arguments doesn't match the number of %-specifiers we are seeing.